femtools/Unittest_tools.F90

```diff
@@ -434,61 +428,18 @@ function fgt(float1, float2, tol) result(greater_than)
 
   end function fgt
 
-  function is_nan_scalar(float1) result(nan)
-    !!< This function checks if float1 is NaN.
-    
-    real, intent(in) :: float1
-    
+  elemental function is_nan(x) result(nan)
+    !! Check if a number is NaN. Safe to use with aggressive optimisations.
+    !!
+    !! @note ieee_is_nan(val) or the older check val /= val are always
+    !! optimised to .false. whenever finite math optimisations are ebaled
+
+    real, intent(in) :: x
     logical :: nan
 
-    nan = ieee_is_nan(float1)
-    
-  end function is_nan_scalar
-  
-  function is_nan_vector(float1) result(nan)
-    !!< This function checks if float1 is NaN.
-    
-    real, dimension(:), intent(in) :: float1
-    
-    logical, dimension(size(float1)) :: nan
-    
-    integer :: i
-    
-    do i = 1, size(float1)
-      nan(i) = is_nan(float1(i))
-    end do
-    
-  end function is_nan_vector
-  
-  function is_nan_tensor2(float1) result(nan)
-    !!< This function checks if float1 is NaN.
-    
-    real, dimension(:, :), intent(in) :: float1
-    
-    logical, dimension(size(float1, 1), size(float1, 2)) :: nan
-    
-    integer :: i
-    
-    do i = 1, size(float1, 1)
-      nan(i, :) = is_nan(float1(i, :))
-    end do
-    
-  end function is_nan_tensor2
-  
-  function is_nan_tensor3(float1) result(nan)
-    !!< This function checks if float1 is NaN.
-    
-    real, dimension(:, :, :), intent(in) :: float1
-    
-    logical, dimension(size(float1, 1), size(float1, 2), size(float1, 3)) :: nan
-    
-    integer :: i
-    
-    do i = 1, size(float1, 1)
-      nan(i, :, :) = is_nan(float1(i, :, :))
-    end do
-    
-  end function is_nan_tensor3
+    nan = (.not. (x <= huge(x) .and. x >= -huge(x))) .and. (.not. abs(x) > huge(x))
```

Comment:
Yes, @gnikit made these changes. Nonetheless, I found those two StackOverflow pages: [1](https://stackoverflow.com/questions/35638400/handling-nan-infinity-and-numbers-larger-than-double-precision-in-fortran) and [2](https://stackoverflow.com/questions/17389958/is-there-a-standard-way-to-check-for-infinite-and-nan-in-fortran-90-95). Do you think there is any good idea in there? If not I could probably just add the first link as a comment as you suggested. Any way to do this without triggering the whole CI again and still being able to merge the PR?

Reply:
Maybe we can add a unittest for the unittest expression to check if it is able to evaluate .true. and .false. with optimisations enabled, but I think it might be an overkill.

>Any way to do this without triggering the whole CI again and still being able to merge the PR?

I don't think it's possible with our current setup. We would have to rewrite the GitHub Actions for that. The only junky solution I could think off is [this](https://stackoverflow.com/questions/70377390/is-there-any-way-to-trigger-a-specific-github-action-workflow-by-commit-message)